zephyr/lib/regions_mm.c

```diff
@@ -216,6 +216,123 @@ struct vmh_heap *vmh_init_heap(const struct vmh_heap_config *cfg,
+ */
+static inline bool vmh_is_region_used(struct sys_mem_blocks *blocks, const uintptr_t ptr,
+				      const size_t size)
+{
+	__ASSERT_NO_MSG(IS_ALIGNED(size, 1 << blocks->info.blk_sz_shift));
+	return !sys_mem_blocks_is_region_free(blocks, UINT_TO_POINTER(ptr),
+					      size >> blocks->info.blk_sz_shift);
+}
+
+/**
+ * @brief Checks whether the memory region is located on memory pages unused by other allocations
+ *
+ * @param allocator Pointer to the sys_mem_blocks allocator
+ * @param ptr Pointer to the memory region to be checked
+ * @param size Size of memory region
+ * @param begin Address to a variable where address of the first unused memory page will be placed
+ * @param out_size Address to a variable where size of unused memory pages will be placed
+ *
+ * @retval true if there is any unused memory page
+ */
+static bool vmh_get_map_region_boundaries(struct sys_mem_blocks *blocks, const void *ptr,
+					  const size_t size, uintptr_t *begin, size_t *out_size)
+{
+	__ASSERT_NO_MSG(blocks);
+	__ASSERT_NO_MSG(begin);
+	__ASSERT_NO_MSG(out_size);
+
+	const size_t block_size = 1 << blocks->info.blk_sz_shift;
+	const size_t size_aligned = ALIGN_UP(size, block_size);
+	uintptr_t addr = ALIGN_DOWN((uintptr_t)ptr, CONFIG_MM_DRV_PAGE_SIZE);
+	uintptr_t addr_end = ALIGN_UP((uintptr_t)ptr + size, CONFIG_MM_DRV_PAGE_SIZE);
+	const uintptr_t size_before = (uintptr_t)ptr - addr;
+	const uintptr_t size_after = addr_end - (uintptr_t)ptr - size_aligned;
+
+	__ASSERT_NO_MSG(IS_ALIGNED(size_before, block_size));
+
+	if (size_before && vmh_is_region_used(blocks, addr, size_before))
+		/* skip first page */
+		addr += CONFIG_MM_DRV_PAGE_SIZE;
+
+	if (size_after && vmh_is_region_used(blocks, POINTER_TO_UINT(ptr) + size_aligned,
+					     size_after))
+		/* skip last page */
+		addr_end -= CONFIG_MM_DRV_PAGE_SIZE;
+
+	if (addr >= addr_end)
+		return false;
+
+	*begin = addr;
+	*out_size = addr_end - addr;
+	return true;
+}
+
+/**
+ * @brief Maps memory pages for a memory region if they have not been previously mapped for other
+ *	  allocations.
+ *
+ * @param allocator Pointer to the sys_mem_blocks allocator
+ * @param ptr Pointer to the memory region
+ * @param size Size of memory region
+ *
+ * @retval 0 on success, error code otherwise.
+ */
+static int vmh_map_region(struct sys_mem_blocks *region, void *ptr, size_t size)
+{
+	const size_t block_size = 1 << region->info.blk_sz_shift;
+	uintptr_t begin;
+	int ret;
+
+	if (block_size >= CONFIG_MM_DRV_PAGE_SIZE) {
+		begin = POINTER_TO_UINT(ptr);
+		size = ALIGN_UP(size, CONFIG_MM_DRV_PAGE_SIZE);
+	} else {
+		if (!vmh_get_map_region_boundaries(region, ptr, size, &begin, &size))
+			return 0;
+	}
+
+	ret = sys_mm_drv_map_region(UINT_TO_POINTER(begin), 0, size, SYS_MM_MEM_PERM_RW);
+
+	/* In case of an error, the pages that were successfully mapped must be manually released */
+	if (ret)
+		sys_mm_drv_unmap_region(UINT_TO_POINTER(begin), size);
```

Comment:
@lyakh The zephyr's default memory region mapping function `sys_mm_drv_simple_map_region` skips pages that could not be mapped and returns only an error code. The situation changes if asserts are active. I don't know if this is a bug or intended design. This would require more consultation with the zephyr team and is beyond the scope of this PR.

Reply:
@ceolin @dcpleung should this be fixed?

Reply:
Could you file an issue so we can track it?

Reply:
> Could you file an issue so we can track it?

@dcpleung https://github.com/zephyrproject-rtos/zephyr/issues/73121